### PR TITLE
Use Explicit Flow Control for Cronvoy.

### PR DESCRIPTION
**Pull request by carloseltuerto**
This is an almost complete rewrite of the Cronvoy main logic (CronetUrlRequest and CronetUploadDataStream). More cleanup will be done with next PRs, and much more documentation will also be added - the logic is quite subtile.

Description: Reimplement Cronvoy to take advantage of the Explicit Flow Control
Risk Level: None: only Cronvoy is changed
Testing: CI - original tests are passing, and also all related `@Ignore`
Docs Changes: N/A
Release Notes: N/A
Signed-off-by: Charles Le Borgne <cleborgne@google.com>

### CODE REVIEW

File: library/common/http/client.cc [hunk 1]
```diff
@@ -80,6 +80,16 @@ void Client::DirectStreamCallbacks::encodeData(Buffer::Instance& data, bool end_
     closeStream();
   }
 
+  if (data.length() != 0 && explicit_flow_control_ && !response_data_) {
```
Comment (alyssawilk) [inline library/common/http/client.cc:83]: can we comment why this move is necessary? Sadly I don't think I grok the race condition here

File: library/common/http/client.cc [hunk 2]
```diff
@@ -80,6 +80,25 @@ void Client::DirectStreamCallbacks::encodeData(Buffer::Instance& data, bool end_
     closeStream();
   }
 
+  // The response_data_ is systematically assigned here because resumeData can
+  // incur an asynchronous callback to sendDataToBridge. It might look like
+  // safe to only set the response_data_ after invoking sendDataToBridge in this
```
Comment (alyssawilk) [inline library/common/http/client.cc:85]: Thanks for explaining this one to me.  I think you can cut this down to just the top two lines as that's sufficient for me at least :-)

**Review by alyssawilk — COMMENTED** [review]
LGTM pending comment update
thanks again for both catching and debugging this one!

**Review by alyssawilk — APPROVED** [review]
Woot!  I'm largely trusting you on the Java changes.  Would it make sense for @colibie to review those as well or you think this is good to go?

**Comment by carloseltuerto** [comment]
> Woot! I'm largely trusting you on the Java changes. Would it make sense for @colibie to review those as well or you think this is good to go?

The sad reality is this PR is very hard to review without hours of explanations, and I am sure that there are subtile bugs.  This is code is still "work in progress". There are many of rough edges left, and I have to add a lot of documentation. This is based on CAS logic - efficient but quite tricky.

**Pull request merged**